Incorrect "extension" of CSS gradient with "longer hue" interpolation into regions beyond the first/last color stop

smfr opened the pull request:
#### ae2fd930ce5bcc6dcbc02c7146471f2c06c6c081
<pre>
Incorrect &quot;extension&quot; of CSS gradient with &quot;longer hue&quot; interpolation into regions beyond the first/last color stop
<a href="https://bugs.webkit.org/show_bug.cgi?id=285382">https://bugs.webkit.org/show_bug.cgi?id=285382</a>
<a href="https://rdar.apple.com/142738948">rdar://142738948</a>

Reviewed by Sam Weinig and Darin Adler.

With `longer hue` CSS gradients, between adjacent color stops sharing a color there is a
full hue cycle. However, this logic also ran for the synthetic 0.0 and 1.0 color stops
that we add to extend the end colors if necessary.

Fix by annotating those end stops as &quot;synthetic&quot;; if the current segment has a synthetic
segment as its start or end, just use that stop&apos;s color.

* LayoutTests/imported/w3c/web-platform-tests/css/css-images/gradient/gradient-single-stop-longer-hue-hsl-002.html: More tolerance.
* LayoutTests/TestExpectations: 18 new passing tests.
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
(WebCore::GradientRendererCG::Shading::shadingFunction):
(WebCore::GradientRendererCG::makeShading const):
* Source/WebCore/platform/graphics/cg/GradientRendererCG.h:
(WebCore::GradientRendererCG::Shading::Data::create):
(WebCore::GradientRendererCG::Shading::Data::firstStopIsSynthetic const):
(WebCore::GradientRendererCG::Shading::Data::lastStopIsSynthetic const):
(WebCore::GradientRendererCG::Shading::Data::Data):

Canonical link: <a href="https://commits.webkit.org/295860@main">https://commits.webkit.org/295860@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1693879cafde6d7b8627970dcba87f6e47e5f26c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106353 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16500 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111553 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56949 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108392 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26768 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34605 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80781 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109357 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21200 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95978 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61109 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14077 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56389 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90543 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14112 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114413 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33491 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89851 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33855 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92207 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89555 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22850 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34442 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12258 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29077 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33416 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38828 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33162 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36515 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34760 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig reviewed (approved):
Nice!

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-images/gradient/gradient-single-stop-longer-hue-hsl-002.html:
```diff
@@ -4,7 +4,7 @@
     <meta charset="utf-8">
     <title>Gradient in HSL space</title>
-    <meta name="fuzzy" content="maxDifference=0-8;totalPixels=0-9050">
+    <meta name="fuzzy" content="maxDifference=0-8;totalPixels=0-16900">
```
Do we need to upstream this?

darinadler commented on Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
```diff
@@ -579,11 +579,18 @@ void GradientRendererCG::Shading::shadingFunction(void* info, const CGFloat* raw
-        makeFromComponents<InterpolationSpaceColorType>(stop0.colorComponents), 1.0f - offset,
-        makeFromComponents<InterpolationSpaceColorType>(stop1.colorComponents), offset);
+    auto interpolatedColor = [&]() {
+        if (stop0.offset == 0.0f && data->firstStopIsSynthetic())
+            return makeFromComponents<InterpolationSpaceColorType>(stop0.colorComponents);
+
+        if (stop1.offset == 1.0f && data->lastStopIsSynthetic())
+            return makeFromComponents<InterpolationSpaceColorType>(stop1.colorComponents);
```
I think this needs a comment. To me, these things are not obvious:

1. why this is needed and doesn’t do the right thing without a special case
2. why it’s important to do this only for _synthetic_ stops
3. why this shouldn’t be done in `interpolateColorComponents` instead of here, maybe the fact that the _synthetic_ state isn’t passed through, maybe the fact that this is needed only for CG
4. why having this special case at exactly 0 and 1 doesn’t create discontinuities in the gradient

I think we should consider saying something brief in the code that might clarify rather than relying solely on test cases and change log.

smfr committed:
Incorrect "extension" of CSS gradient with "longer hue" interpolation into regions beyond the first/last color stop
https://bugs.webkit.org/show_bug.cgi?id=285382
rdar://142738948

Reviewed by Sam Weinig and Darin Adler.

With `longer hue` CSS gradients, between adjacent color stops sharing a color there is a
full hue cycle. However, this logic also ran for the synthetic 0.0 and 1.0 color stops
that we add to extend the end colors if necessary.

Fix by annotating those end stops as "synthetic"; if the current segment has a synthetic
segment as its start or end, just use that stop's color.

* LayoutTests/imported/w3c/web-platform-tests/css/css-images/gradient/gradient-single-stop-longer-hue-hsl-002.html: More tolerance.
* LayoutTests/TestExpectations: 18 new passing tests.
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
(WebCore::GradientRendererCG::Shading::shadingFunction):
(WebCore::GradientRendererCG::makeShading const):
* Source/WebCore/platform/graphics/cg/GradientRendererCG.h:
(WebCore::GradientRendererCG::Shading::Data::create):
(WebCore::GradientRendererCG::Shading::Data::firstStopIsSynthetic const):
(WebCore::GradientRendererCG::Shading::Data::lastStopIsSynthetic const):
(WebCore::GradientRendererCG::Shading::Data::Data):

Canonical link: https://commits.webkit.org/295860@main
